File: grail/versions.py
"""Version ordering with the exact semantics of Nix's builtins.compareVersions.

The reference is nextComponent/componentsLT in Nix's names.cc: a version
splits into components that are maximal runs of digits or maximal runs of
non-digits, with '.' and '-' acting as separators that appear in no
component. Ordering rules, in the order componentsLT checks them:

  - two numeric components compare as integers
  - an empty component is smaller than a numeric one
  - "pre" is smaller than anything except "pre"
  - a numeric component is larger than any non-numeric one
  - anything else compares as a plain string (so "" < "rc": only pre is
    special, an -rc suffix sorts AFTER the release)
"""

from __future__ import annotations

# The separator characters nextComponent skips over.
_SEPARATORS = ".-"

# The one magic component that sorts before the release it suffixes.
_PRE = "pre"


def components(version: str) -> list[str]:
    """Split a version string the way Nix's nextComponent does."""
    parts: list[str] = []
    i = 0
    n = len(version)
    while i < n:
        # skip separators between components
        if version[i] in _SEPARATORS:
            i += 1
            continue

        # a component is a run of digits, or a run of everything else
        start = i
        if version[i].isdigit():
            while i < n and version[i].isdigit():
                i += 1
        else:
            while i < n and not version[i].isdigit() and version[i] not in _SEPARATORS:
                i += 1
        parts.append(version[start:i])
    return parts


def _component_less(a: str, b: str) -> bool:
    """componentsLT from names.cc, verbatim in Python."""
    a_num = int(a) if a.isdigit() else None
    b_num = int(b) if b.isdigit() else None

    if a_num is not None and b_num is not None:
        return a_num < b_num
    if a == "" and b_num is not None:
        return True
    if a == _PRE and b != _PRE:
        return True
    if b == _PRE:
        return False
    if b_num is not None:
        return True
    if a_num is not None:
        return False
    return a < b


def compare(a: str, b: str) -> int:
    """Return <0, 0 or >0 exactly as builtins.compareVersions would."""
    ca = components(a)
    cb = components(b)

    # walk both component lists, padding the shorter with ""
    for i in range(max(len(ca), len(cb))):
        x = ca[i] if i < len(ca) else ""
        y = cb[i] if i < len(cb) else ""
        if x == y:
            continue
        if _component_less(x, y):
            return -1
        if _component_less(y, x):
            return 1
        # components differ as strings but neither is less: treat as equal
        # (cannot happen with componentsLT's total order, kept for safety)
    return 0


def sort_key(version: str):
    """A sort key consistent with compare(), for ranking versions."""
    import functools

    return functools.cmp_to_key(compare)(version)

File: site/js/versions.js
// Nix's builtins.compareVersions, ported byte-for-byte from grail's Python
// port of names.cc (which tests/test_versions.py holds against real nix).
// The quirks are load-bearing: only "pre" sorts before its release, an
// empty component is below a number but above nothing else, letters sort
// below numbers.

const SEPARATORS = new Set([".", "-"]);
const PRE = "pre";

export function components(version) {
  const parts = [];
  let i = 0;
  const n = version.length;
  const isDigit = (c) => c >= "0" && c <= "9";
  while (i < n) {
    if (SEPARATORS.has(version[i])) {
      i += 1;
      continue;
    }
    const start = i;
    if (isDigit(version[i])) {
      while (i < n && isDigit(version[i])) i += 1;
    } else {
      while (i < n && !isDigit(version[i]) && !SEPARATORS.has(version[i])) i += 1;
    }
    parts.push(version.slice(start, i));
  }
  return parts;
}

function componentLess(a, b) {
  const aNum = /^[0-9]+$/.test(a) ? BigInt(a) : null;
  const bNum = /^[0-9]+$/.test(b) ? BigInt(b) : null;
  if (aNum !== null && bNum !== null) return aNum < bNum;
  if (a === "" && bNum !== null) return true;
  if (a === PRE && b !== PRE) return true;
  if (b === PRE) return false;
  if (bNum !== null) return true;
  if (aNum !== null) return false;
  return a < b;
}

export function compare(a, b) {
  const ca = components(a);
  const cb = components(b);
  const len = Math.max(ca.length, cb.length);
  for (let i = 0; i < len; i += 1) {
    const x = i < ca.length ? ca[i] : "";
    const y = i < cb.length ? cb[i] : "";
    if (x === y) continue;
    if (componentLess(x, y)) return -1;
    if (componentLess(y, x)) return 1;
  }
  return 0;
}

// component-wise prefix: 3.8 accepts 3.8.9 and 3.8, refuses 3.81
export function isPrefix(pattern, candidate) {
  const p = components(pattern);
  const c = components(candidate);
  if (p.length > c.length) return false;
  return p.every((part, i) => part === c[i]);
}
